riscv/src/code_gen_16.rs

```diff
@@ -0,0 +1,1805 @@
+    instr branch_if_greater_or_equal_signed XL, YL, l: label
+        link ~> (tmp1_h, tmp1_l) = regs.mload(XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(YL, STEP + 1)
+        link ~> wrap_bit_3 = arith_bb.cmp(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+    {
+        tmp1_h = X_b1 + Y_7bit * 0x100 + wrap_bit * 0x8000,
+        tmp2_h = X_b2 + Y_7bit_2 * 0x100 + wrap_bit_2 * 0x8000,
+
+        Y_b5 = (1 - wrap_bit) * (1 - wrap_bit_2),
+        Y_b6 = wrap_bit * wrap_bit_2,
+        Y_b7 = (1 - wrap_bit) * wrap_bit_2,
+
+        //Y_b8 = wrap_bit * (1 - wrap_bit_2),
+        X_b3 = Y_b5 * (1 - wrap_bit_3) +
+               Y_b6 * (1 - wrap_bit_3) +
+               Y_b7, /* * 1 */
+               /* Y_b8 * 0, */
+
+        pc' = X_b3 * l + (1 - X_b3) * (pc + 1)
+    }
+
+    // Stores 1 in register W if val(X) >= val(Y) <=> not(val(X) < val(Y)).
+    instr is_greater_or_equal XL, YL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(YL, STEP + 1)
+        link ~> wrap_bit = arith_bb.cmp(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link ~> regs.mstore(WL, STEP + 2, 0, 1 - wrap_bit);
+
+    // Branches to `l` if val(X) >= val(Y) <=> not(val(X) < val(Y)).
+    instr is_greater_or_equal_signed XL, YL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(YL, STEP + 1)
+        link ~> wrap_bit_3 = arith_bb.cmp(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link ~> regs.mstore(WL, STEP + 2, 0, X_b3)
+    {
+        tmp1_h = X_b1 + Y_7bit * 0x100 + wrap_bit * 0x8000,
+        tmp2_h = X_b2 + Y_7bit_2 * 0x100 + wrap_bit_2 * 0x8000,
+
+        Y_b5 = (1 - wrap_bit) * (1 - wrap_bit_2),
+        Y_b6 = wrap_bit * wrap_bit_2,
+        Y_b7 = (1 - wrap_bit) * wrap_bit_2,
+
+        //Y_b8 = wrap_bit * (1 - wrap_bit_2),
+        X_b3 = Y_b5 * (1 - wrap_bit_3) +
+               Y_b6 * (1 - wrap_bit_3) +
+               Y_b7 /* * 1 */
+               /* Y_b8 * 0, */
+    }
+
+    // Stores val(X) * Z + W in register Y.
+    instr affine XL, YL, ZH, ZL, WH, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(XL, STEP)
+        link ~> (tmp3_h, tmp3_l, tmp2_h, tmp2_l) = arith_mul.mul(tmp1_h, tmp1_l, ZH, ZL)
+        // we ignore tmp3 because that's the high 32 bits of the 64 bits multiplication result
+        link ~> (tmp4_h, tmp4_l) = arith_bb.add(tmp2_h, tmp2_l, WH, WL)
+        link ~> regs.mstore(YL, STEP + 1, tmp4_h, tmp4_l);
+
+    // ================= wrapping instructions =================
+
+    // Computes V = val(X) + val(Y) + Z, wraps it in 32 bits, and stores the result in register W.
+    // Requires 0 <= V < 2**33.
+    instr add_wrap XL, YL, ZH, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(YL, STEP + 1)
+        link ~> (tmp3_h, tmp3_l) = arith_bb.add(tmp2_h, tmp2_l, ZH, ZL)
+        link ~> (tmp4_h, tmp4_l) = arith_bb.add(tmp1_h, tmp1_l, tmp3_h, tmp3_l)
```

Comment:
We need to check specifically, but sometimes negative numbers are passed here, in which case we do need both limbs and proper machine arithmetic instead of field.